pkg/receive/tsdb.go

```diff
@@ -0,0 +1,132 @@
+package receive
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	promtsdb "github.com/prometheus/prometheus/tsdb"
+)
+
+type FlushableStorage struct {
+	*promtsdb.DB
+
+	path string
+	l    log.Logger
+	r    prometheus.Registerer
+	opts *tsdb.Options
+
+	stopped bool
+	mu      sync.Mutex
+}
+
+// NewFlushableStorage returns a new storage backed by a TSDB database that is configured for Prometheus.
+func NewFlushableStorage(path string, l log.Logger, r prometheus.Registerer, opts *tsdb.Options) *FlushableStorage {
+	return &FlushableStorage{
+		path:    path,
+		l:       l,
+		r:       r,
+		opts:    opts,
+		stopped: true,
+	}
+}
+
+// Get returns a reference to the underlying storage.
+func (f *FlushableStorage) Get() *promtsdb.DB {
+	return f.DB
+}
+
+// Open starts the TSDB.
+func (f *FlushableStorage) Open() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if !f.stopped {
+		return nil
+	}
+	db, err := tsdb.Open(
+		f.path,
+		log.With(f.l, "component", "tsdb"),
+		&UnRegisterer{f.r},
+		f.opts,
+	)
+	f.DB = db
+	f.stopped = false
+	return err
+}
+
+// Flush stops the storage and flushes the WAL to blocks.
+// Note: this operation results in a closed storage.
+func (f *FlushableStorage) Flush() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.stopped {
+		return nil
+	}
+	tmp := filepath.Join(f.Dir(), "tmp")
+	if err := os.Mkdir(tmp, os.ModePerm); err != nil && !os.IsExist(err) {
+		return errors.Wrap(err, "making tmp dir")
+	}
+	defer os.RemoveAll(tmp)
+	f.DisableCompactions()
```

Comment:
logically this is the right place, but `DisableCompactions` has no effect on started compactions.
For example a call to `DisableCompactions` in the middle of a running compaction will not stop it so a call to `Snapshot` would still happen in the middle of a running compaction.

Reply:
Ack, we’ll switch over to the DBReadOnly implementation to make this work correctly :+1:

Reply:
Yes `DBReadOnly` would the safest approach in the long run, but in this case it might also work if you disable the compaction as soon as you open the db.

Reply:
xref https://github.com/prometheus/prometheus/pull/6006